SAMtags.tex

```diff
@@ -453,6 +453,161 @@ \subsubsection{Color space}
+N & N &       & Ambiguity code; any mod & \\
+\end{tabular}
+\end{center}
+
+% MP was the former quality score for MM.  However being Phred scores
+% it can only reasonable record probabilities for highly likely
+% events, making it inappropriate for callers (eg ONT's) that wish to
+% jointly call probabilities for the entire trained set of
+% possibilities.  We could use log-odds, similar to how early Illumina
+% runs did to record likelihoods for A, C, G and T irrespective of
+% call, but for now we're using linear-scaled probabilities.  These
+% are in the ML tag.
+%
+% The MP tag is left here for now as the jury is still out on whether
+% we'll need it in the future.
+%
+% \item[MP:Z:\tagvalue{qualities}]
+% \hfill\\
+% The optional {\tt MP} tag lists the Phred qualities of each modification listed in the {\tt MM} tag in the order they occur.
+% The qualities are encoded in the same manner as the primary {\sf QUAL} field; one byte per quality with ASCII value Phred score + 33.
+% A space character (`{\tt \textvisiblespace}') should be used as a separator between concatenated quality strings when multiple modification lists are present in the {\tt MM} tag.
+% The length should match the number of position deltas from {\tt MM} plus 1 per space character required.
+%
+% For example ``{\tt MM:Z:C+m,5,12,3;C+h,57;}'' may have an associated
+% quality tag of ``{\tt MP:Z:5EB /}''.
+%
+% Where multiple modification types are listed together, such as in ``{\tt MM:Z:C+mh,5,12,3;}'' the quality values are interleaved in order ({\tt m} at 6, {\tt h} at 6, {\tt m} at 19, {\tt h} at 19 and so on), giving 6 quality values in total for this example.
+%
+% Quality values for ambiguity codes give the likelihood that the
+% modification is one of the possible codes compatible with that
+% ambiguity code.  For example {\tt MM:Z:C+C,10; MP:Z:+} indicates a C
+% call with an unspecified modification and the phred score of 10 (ASCII
+% value {\tt +}).  This corresponds to a 90\% chance of the base being
+% modified.
+%
+% To represent several possible modifications at the same site the {\tt MP} tag can be used to indicate the probabilities of each possibility.
+% The values used should be absolute probabilities, not relative between the alternatives.
+% For example, a C base that has 95\% chance of being modified with 5mC being three times more likely than 5hmC will encode 5mC with 67.5\% probability ($0.9 * 0.75$ giving phred score 5, ASCII value {\tt \&})and 5hmC with 22.5\% probability ($0.9 * 0.25$ giving phred score 1, ASCII value {\tt "}).
+% This could be represented with ``{\tt MM:Z:C+m,10;C+h,10; MP:Z:" \&}''.
+
+\item[Ml:B:C,\tagvalue{scaled-probabilities}]
+\hfill\\
+The optional {\tt Ml} tag lists the probability of each modification listed in the {\tt Mm} tag being correct, in the order that they occur.
+The probabilities are scaled to fit in one byte such that $p=0.0$ is value 0 and $p=1.0$ is value 255.
```

Comment:
Very minor point, but for the scaled probabilities it seems logical that each integer value represents an equal proportion of the input space (0-1). Thus the value 0 would represent probabilities in the range `0/256` to `1/256` and `255` value would represent the range `255/256` to `256/256==1.0`. Is this worth specifying explicitly?

Reply:
So with that in mind, and your comment below which I now understand, the integer value is centred on the bin.  Thus 0 on average (assuming uniform distribution, which obviously isn't likely to be true but let's let that slide) will mean p=0.5/256 and 255 means p=255.5/256?

To figure out if this is consistent, I thought about the phred scale definition where Q20 means 0.01 chance of an error (p=0.99).  If we compute an actual p=0.9895 giving phred Q19.79, then we'd be expected to round that to Q20 not Q19.  Thus similarly we're saying although Q20 is 0.01 chance of error, actually anywhere from P(err)=0.0090 to 0.0112 also give Q20.  I think the difference in language is we define phred scales in encoding (probability to integer) whereas the above definition is defined in the reverse (integer to probability).  We never explicitly state though with phred which way rounding is meant to go, and I expect some do floor while others do round (nearest).

Ie after some mental gymnastics, I agree with your assessment.  I do wonder if it's the right way to approach it though.

An alternative would be to define the integer value as being `p*256`, rounding to nearest integer and with an explicit upper limit of 255.  I think that's better than p*255 to avoid the overflow.

Reply:
I'm happy to leave this unspecified, but it seems to my mind that the most logical thing to do would be to have each integer encoding represent an equal range of the probability space. By not including this in the specification, alternative interpretations are likely to result. The issue with the `p*256` rounded and then explicitly clipping to `255` is that the value `0` represents `1/512` stretch of input probability, while `255` represents `3/512` stretch of the input probability space, while all other encodings represent `1/265`.

Reply:
I do think my formula below is equivalent to `p*256` then floor then clip to `255`.

Reply:
The only problem with `p*256` instead of `p*255` is it removes the possibility of stating a certainty.  Ie this base *is* modified.  However I don't see that as a problem given the current (and likely future) technologies.

A value of 255 still corresponds to p=0.996 or Q24 in Phred Scale.  If we ever get technologies able to cope assign modification likelihoods higher than that for individual reads (not consensii) then we can introduce an MP tag to use a different scale, as originally proposed in this PR.

I'll change this PR to use p*256.